app/src/main/java/org/wikipedia/util/ShareUtil.kt

```diff
@@ -0,0 +1,221 @@
+    }
+
+    private fun getClearShareFolder(context: Context): File? {
+        try {
+            val dir = File(getShareFolder(context), "share")
+            FileUtil.deleteRecursively(dir)
+            return dir
+        } catch (caught: Throwable) {
+            e("Caught " + caught.message, caught)
+        }
+        return null
+    }
+
+    private fun getShareFolder(context: Context): File {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) context.cacheDir
+        else context.getExternalFilesDir(null)!!
+    }
+
+    private fun cleanFileName(fileName: String): String {
+        // Google+ doesn't like file names that have characters %28, %29, %2C
+        var fileName = fileName
+        fileName = fileName.replace("%2[0-9A-F]".toRegex(), "_")
+                .replace("[^0-9a-zA-Z-_\\.]".toRegex(), "_")
+                .replace("_+".toRegex(), "_")
+        // ensure file name ends with .jpg
+        if (!fileName.endsWith(".jpg")) {
+            fileName = "$fileName.jpg"
+        }
+        return fileName
+    }
+
+    @JvmStatic
+    fun createChooserIntent(targetIntent: Intent, context: Context): Intent {
+        val chooser = Intent.createChooser(targetIntent, null)
+        var intents = queryIntents(context, targetIntent, false)
+        if (intents.isEmpty()) {
+            // This implies that the Wikipedia app itself has been chosen as the default handler
+            // for our links, so we need to explicitly build a chooser that contains other activities.
+            intents = queryIntents(context, targetIntent, true)
+            if (!intents.isEmpty()) {
+                chooser.putExtra(Intent.EXTRA_INITIAL_INTENTS, intents.toTypedArray<Parcelable?>())
+            }
+        }
+        return chooser
+    }
+
+    private fun queryIntents(context: Context, targetIntent: Intent, replaceUri: Boolean): List<Intent?> {
+        val intents: MutableList<Intent?> = ArrayList()
+        var queryIntent = targetIntent
+        if (replaceUri) {
+            queryIntent = Intent(targetIntent)
+            queryIntent.data = Uri.parse("https://example.com/")
+        }
+        for (intentActivity in context.packageManager.queryIntentActivities(queryIntent, 0)) {
+            if (!isIntentActivityBlacklisted(intentActivity, APP_PACKAGE_REGEX)) {
+                intents.add(buildLabeledIntent(targetIntent, intentActivity))
+            }
+        }
+        return intents
+    }
+
+    @JvmStatic
+    fun canOpenUrlInApp(context: Context, url: String): Boolean {
+        var canOpen = false
+        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
+        for (intentActivity in context.packageManager.queryIntentActivities(intent, 0)) {
+            if (getPackageName(intentActivity).matches(APP_PACKAGE_REGEX.toRegex())) {
+                canOpen = true
+                break
+            }
+        }
```

Comment:
Didn't verify but looks like this can be simplified as following:
`canOpen = context.packageManager.queryIntentActivities(intent, 0).any { getPackageName(it).matches(APP_PACKAGE_REGEX.toRegex()) }`.